### PR TITLE
Feat: Agent API Origin Whitelisting

### PR DESCRIPTION
**Pull request by ardafincan**
- **What kind of change does this PR introduce?**

This PR introduces agent API Origin Whitelisting feature for security purposes. User can define allowed origins for accessing agents via API calls.

- **Why was this change needed?** 
Issue #1984 explains the need of this feature, it can be summarized as security improvement.

### CODE REVIEW

**Comment by ardafincan** [comment]
Hey @dartpain, you have any time to review this one?

File: application/api/answer/routes/base.py [hunk 1]
```diff
@@ -123,6 +123,37 @@ def check_usage(
...
+        """
+        api_key = agent_config.get("user_api_key")
+        if not api_key:
+            return None
+        
+        agents_collection = self.db["agents"]
+        agent = agents_collection.find_one({"key": api_key})
+        
+        request_origin = request.headers.get("Origin")
+        granted_origins = json.loads(agent.get("granted_origins", "[]"))
```
Comment (siiddhantt) [inline application/api/answer/routes/base.py]: This would crash with `AttributeError` if agent is not found, could add a null check like in `check_usage()`

File: frontend/src/agents/NewAgent.tsx [hunk 2]
```diff
@@ -1058,6 +1062,118 @@ export default function NewAgent({ mode }: { mode: 'new' | 'edit' | 'draft' }) {
...
+                                ],
+                              });
+                              setOriginInput('');
+                            }
+                          }
+                        }}
+                      />
+                      <button
+                        type="button"
+                        onClick={() => {
```
Comment (siiddhantt) [inline frontend/src/agents/NewAgent.tsx]: Consider refactoring this into a shared `handleAddOrigin()` as there is duplicated code just before.

File: application/api/user/agents/routes.py [hunk 3]
```diff
@@ -362,6 +369,7 @@ def post(self):
                 "image": image_url,
                 "source": source_field,
                 "sources": sources_list,
+                "granted_origins": data.get("granted_origins", []),
```
Comment (siiddhantt) [inline application/api/user/agents/routes.py]: Should parse the JSON string received from frontend for consistency. Currently storing JSON string, then parsing it later on retrieval - better to normalize on write.

File: application/api/user/agents/routes.py
Comment (siiddhantt) [inline application/api/user/agents/routes.py]: Add granted_origins to early JSON parsing (like sources/tools):

```bash
if "granted_origins" in data:
    try:
        data["granted_origins"] = json.loads(data["granted_origins"])
    except json.JSONDecodeError:
        data["granted_origins"] = []
```

Comment (siiddhantt) [inline application/api/user/agents/routes.py]: Add "granted_origins" to json_fields list for early parsing:

`json_fields = ["tools", "sources", "json_schema", "granted_origins"]`

This ensures consistency with how other JSON fields are handled.

File: frontend/src/agents/NewAgent.tsx [hunk 4]
```diff
@@ -1058,6 +1062,118 @@ export default function NewAgent({ mode }: { mode: 'new' | 'edit' | 'draft' }) {
...
+                      >
+                        Add
+                      </button>
+                    </div>
+                    {agent.granted_origins &&
+                      agent.granted_origins.length > 0 && (
+                        <div className="space-y-2">
+                          {agent.granted_origins.map((origin, index) => (
+                            <div
+                              key={index}
```
Comment (siiddhantt) [inline frontend/src/agents/NewAgent.tsx]: Could improve it by using a unique key instead of index keys, maybe `key={origin}`?

**Comment by siiddhantt** [comment]
Hi @ardafincan thanks for the PR! 
Could you also add tests to improve coverage and verify implementation. Tests like `Valid/invalid origin checks`, `Missing Origin header`, `Invalid API key`, `Multiple origins`, `URL normalization edge cases`  would significantly improve coverage!

**Comment by ardafincan** [comment]
Hey @siiddhantt, thanks for your detailed review. I am a bit busy this week, but I will handle the changes and test additions you wanted as soon as possible.

**Comment by siiddhantt** [comment]
No worries @ardafincan thanks for contributing!

**Comment by siiddhantt** [comment]
Hi @ardafincan any update on this?
